docs: add Understanding Kubernetes Load balancing pillar page (#962)
cilium/cilium.io: src/posts/2026-04-25-understanding-kubernetes-load-balancing/index.md

While kube-proxy is limited to Layer 4 (IP/Port), Cilium can perform Layer 7 (HTTP/gRPC) load balancing and observability, providing a more identity-aware networking layer that is both faster and more resilient at scale.

![Internal Load Balancing](./images/internal-load-balancing.png)

### External Load Balancing

External load balancing is the gateway that connects outside users to services running inside the cluster. While internal load balancing manages East-West traffic (pod-to-pod), external load balancing handles North-South traffic (internet-to-pod).

This is achieved through three primary methods:

- NodePort, which exposes a specific port on every node's IP.
- LoadBalancer, which integrates with infrastructure providers.
- Ingress, which acts as a Layer 7 smart router (handling hostnames and SSL), sits behind one of the aforementioned service types.

> - Gateway API is a family of API kinds that provide dynamic infrastructure provisioning and advanced traffic routing. Gateway API is the successor to Ingress. Cilium has native Gateway API support built directly into its eBPF datapath; no separate ingress controller is needed. A Gateway resource backed by Cilium gets L7 routing, native load balancing, and full Hubble observability on every request without an additional proxy in the path.

Reviewer: ```suggestion
- Gateway API is a family of API kinds that provide dynamic infrastructure provisioning and advanced traffic routing. Gateway API is the successor to Ingress. Cilium has native Gateway API support built directly into its Envoy and eBPF data path; no separate ingress controller is needed. A Gateway resource backed by Cilium gets L7 routing, native load balancing, and full Hubble observability on every request without an additional proxy in the path.
```

---
path: /blog/2026/04/25/understanding-kubernetes-load-balancing
date: 2026-04-25T09:13:00.000Z

> title: 'Understanding Kubernetes Load balancing'

Reviewer: ```suggestion
title: 'Understanding Kubernetes Load Balancing'
```

isFeatured: false
ogImage: images/understanding-kubernetes-load-balancing-cover.png
ogSummary: 'Explore the Kubernetes load balancing model. This comprehensive guide describing Services, Ingress, north-south vs. east-west traffic, and how eBPF & Cilium enables scalable, high-performance kube-proxy replacement.'

categories:
  - Technology
tags:
  - Cilium
  - Load Balancing
  - XDP
  - LB
  - BGP
  - L2 Aware
  - ARP
  - CNI
  - Services
  - L4LB
  - Kubernetes
---

import authors from 'utils/author-data';

# Understanding Kubernetes Load Balancing

# I.Introduction

Since pods are ephemeral, constantly changing, replaced, scaled up, scaled down, and, the IP addresses change. To ensure proper utilization and desired performance, each healthy running instance of your application has to get a portion of the traffic. Traffic has to be distributed by load balancing between these instances; the distribution weight will depend on the algorithm.

> **Kubernetes Load Balancing** is a means to distribute network traffic to application instances. This can be within the same cluster or different compute regions or clusters.

Reviewer: ```suggestion
Kubernetes Load Balancing is a means to distribute network traffic to application instances. This can be within the same cluster or different compute regions or clusters.

To ensure proper utilization and desired performance, each healthy running instance of your application has to get a portion of the traffic. Traffic has to be distributed by load balancing between these instances; the distribution weight will depend on the algorithm.
```

## Services in Kubernetes

Services are used to expose applications running inside the cluster behind a single outward-facing endpoint, even when the workload is split across multiple backends. This is what makes load balancing possible.

![Load Balancer](./images/image1.png)

## Why Kubernetes Load Balancing?

The main goals that Load Balancing is trying to achieve are Service Reliability, Availability, and Performance (Horizontal Scalability).
Load balancing ensures reliability by using health checks and readiness probes to steer traffic away from failing pods. By distributing incoming requests across multiple replicas, it guarantees high availability, ensuring that the failure of a single pod or even an entire physical node does not result in application downtime.

# II.How does Kubernetes Load Balancing Work?

> Speaking of Load Balancing in Kubernetes, clusters can be split into two perspectives.

Reviewer: ```suggestion
Load Balancing in Kubernetes clusters can be split into two perspectives.
```

### Internal Load Balancing

Internal load balancing refers to the distribution of traffic within a Kubernetes cluster itself among pods of the same application or service.
Kubernetes uses services to implement internal load balancing. For internal load balancing services with designated cluster IPs (reachable within the cluster), fit the purpose.

When pod A needs to talk to pod B inside the same cluster, it doesn’t use pod B’s direct IP; instead, it uses a Kubernetes service (stable virtual IP).
By default, Virtual IPs are managed by kube-proxy, a network agent running on each node. Kube-proxy watches the Kubernetes API for changes to Services and Endpoints and translates them into local networking rules.

Originally, this was done using iptables, which relies on sequential list processing, or IPVS, which improved performance through hash tables and advanced algorithms like Round-robin and Least Connection.

> However, as clusters scale, the overhead of managing thousands of iptables rules can degrade performance. To improve performance, Cilium can replace kube-proxy and its iptables. By using eBPF, Cilium processes packets at the lowest level of the network stack without the context switching required by iptables. This provides significantly higher throughput, lower latency, and more granular security (Network Policies).

Reviewer: ```suggestion
However, as clusters scale, the overhead of managing thousands of iptables rules can degrade performance. To improve performance, Cilium can replace kube-proxy and iptables. By using eBPF, Cilium processes packets at the lowest level of the network stack without the context switching required by iptables. This provides significantly higher throughput, lower latency, and more granular security.
```

While kube-proxy is limited to Layer 4 (IP/Port), Cilium can perform Layer 7 (HTTP/gRPC) load balancing and observability, providing a more identity-aware networking layer that is both faster and more resilient at scale.

> ![Internal Load Balancing](./images/image2.png)

Reviewer: this image isn't clear to me

Author: changed the image

### External Load Balancing

External load balancing is the gateway that connects outside users to services running inside the cluster. While internal load balancing manages East-West traffic (pod-to-pod), external load balancing handles North-South traffic (internet-to-pod).

This is achieved through three primary methods:

- NodePort, which exposes a specific port on every node's IP.
- LoadBalancer, which integrates with infrastructure providers.

> - Ingress, which acts as a Layer 7 smart router (handling hostnames and SSL), sits behind one of the aforementioned service types.

Reviewer: should gateway api also be added here?

Author: Added gateway API in the section and a diagram

L2-Awareness allows Kubernetes services to be reachable via Address Resolution Protocol (ARP) for IPv4 or Neighbor Discovery Protocol (NDP) for IPv6. Using a feature like Cilium L2 Announcements, a leader node responds to network queries by broadcasting its own MAC address as the destination for a Service's Virtual IP. This bridges the gap between the physical switch and the virtual cluster, making services visible on the local area network (LAN) without requiring complex routing protocols like BGP.

Before Kubernetes can distribute traffic to pods, the external network must first know which physical node in the cluster owns the Service IP. In cloud environments, this is handled by the provider’s Software Defined Network. However, in on-premises or bare-metal environments, the cluster must manage its own physical identity.

#### L2-Aware Load Balancing and Service Announcement in Cilium

An L2-aware load balancer is a networking solution designed to bridge the gap between a physical network (Layer 2 of the OSI model) and a virtualized environment like Kubernetes.

> Standard load balancers usually operate at Layer 4 (IPs and Ports) or Layer 7 (URLs and Cookies), and an L2-aware load balancer handles the first mile of a packet's journey, finding the physical machine that owns a specific Virtual IP.

Reviewer: ```suggestion
```

1. XDP Hook: The eBPF program attached to the NIC intercepts the packet.
2. Maglev Lookup: Cilium uses a consistent hashing algorithm (Maglev) to select a backend Pod, ensuring that traffic from the same flow always hits the same destination.
   Kubernetes service load balancing implemented by Cilium or Kube-proxy selects backends randomly and ensures that the traffic remains sticky to that backend. In the case of node failure, the upstream load balancer selects a different load balancing node that has no context of the failed node. This can lead to an unexpected disruption on connection oriented protocols.
   Maglev consistent hashing minimizes such disruption by ensuring each load-balancing node has a consistent view and ordering of the backend lookup table.
3. Encapsulation/DSR: The packet is either encapsulated (VXLAN/Geneve) or sent via Direct Server Return (DSR) to the target node.
4. XDP_TX: The modified packet is sent back out the same interface immediately, completely bypassing the host's networking stack. XDP_TX is an XDP action that involves TX bouncing the received packet page back out the same NIC it arrived on.

> Read More: [https://cilium.io/blog/2020/11/10/cilium-19/](https://cilium.io/blog/2020/11/10/cilium-19/)

Reviewer: I would make this link more specific

Author: Changed to a specific post link

## Operating Modes

> L4 LoadBalancers work in two modes:

Reviewer: ```suggestion
L4 Load Balancers work in two modes:
```

# IV. Summary

Kubernetes load balancing is a critical architecture for managing traffic across ephemeral, scaled application instances. By decoupling fixed service endpoints from transient pod IP addresses, it ensures Service Reliability, High Availability, and Performance.

To achieve production-grade resilience, these layers use health checks and automated failover mechanisms. Whether through traditional proxy modes or high-performance passthrough with Direct Server Return (DSR) and Maglev consistent hashing, Kubernetes load balancing ensures that network traffic is distributed predictably and efficiently across the entire cluster lifecycle.

Cilium redefines this stack by replacing legacy components like kube-proxy and iptables with a unified, eBPF-native data path. This shift provides three distinct advantages across the networking layers **Foundation (L2/L3)** Cilium bridges the gap for on-premises and bare-metal environments, **Performance (L4/XDP)** By leveraging XDP (eXpress Data Path), Cilium processes packets at the earliest possible point in the network driver and **Intelligence (L7/Ingress)**, At the application level, Cilium provides identity-aware routing.

> <BlogAuthor {...authors.CharityMbisi} />

Reviewer: this isn't working
